[Fix #6115] Fix a false positive for `Lint/OrderedMagicComments`

koic opened the pull request:
Fixes #6115.

This PR fixes a false positive for `Lint/OrderedMagicComments` when using `encoding: Encoding::SJIS` Hash notation after `frozen_string_literal` magic comment.

There was a false positive to misunderstanding the following Hash notation as an encoding magic comment.

```ruby
# frozen_string_literal: true

x = { encoding: Encoding::SJIS }
puts x
```

This PR changes the pattern match of the magic encoding comment so that beginning of line always starts with `#`.

-----------------

Before submitting the PR make sure the following are checked:

* [x] Wrote [good commit messages][1].
* [x] Commit message starts with `[Fix #issue-number]` (if the related issue exists).
* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [x] Added tests.
* [x] Added an entry to the [Changelog](../blob/master/CHANGELOG.md) if the new code introduces user-observable changes. See [changelog entry format](../blob/master/CONTRIBUTING.md#changelog-entry-format).
* [x] The PR relates to *only* one subject with a clear title
  and description in grammatically correct, complete sentences.
* [x] Run `rake default` or `rake parallel`. It executes all tests and RuboCop for itself, and generates the documentation.

[1]: http://chris.beams.io/posts/git-commit/

bbatsov commented on lib/rubocop/magic_comment.rb:
```diff
@@ -190,7 +190,7 @@ def frozen_string_literal; end
       # Match `encoding` or `coding`
       def encoding
-        extract(/\#* \b(?:en)?coding: (#{TOKEN})/i)
+        extract(/\A\#.*\b(?:en)?coding: (#{TOKEN})/i)
```
I don't think `\A` here is a good idea. If a comment is indented this won't match. That's probably bad style, but I think it will work just fine.

koic replied:
I see. I added `\A` just like the pattern match for `frozen_string_literal` magic comment below.

```ruby
def extract_frozen_string_literal
  extract(/\A#\s*frozen[_-]string[_-]literal:\s*(#{TOKEN})\s*\z/i)
end
```

https://github.com/rubocop-hq/rubocop/pull/6117/files#diff-6e19cd40a821113e5df8e1f60f634e7bR206

However, this is not good when indented as pointed out.

For both pattern matching, I'd like to change from `A\#` to `\A\s*\#` to accept indentation.

```diff
% g diff
diff --git a/lib/rubocop/magic_comment.rb b/lib/rubocop/magic_comment.rb
index 60d727c49..8c5a17dcb 100644
--- a/lib/rubocop/magic_comment.rb
+++ b/lib/rubocop/magic_comment.rb
@@ -190,7 +190,7 @@ module RuboCop
     class SimpleComment < MagicComment
       # Match `encoding` or `coding`
       def encoding
-        extract(/\A\#.*\b(?:en)?coding: (#{TOKEN})/i)
+        extract(/\A\s*\#.*\b(?:en)?coding: (#{TOKEN})/i)
       end

       private
@@ -203,7 +203,7 @@ module RuboCop
       # Case-insensitive and dashes/underscores are acceptable.
       # @see https://git.io/vM7Mg
       def extract_frozen_string_literal
-        extract(/\A#\s*frozen[_-]string[_-]literal:\s*(#{TOKEN})\s*\z/i)
+        extract(/\A\s*#\s*frozen[_-]string[_-]literal:\s*(#{TOKEN})\s*\z/i)
       end
     end
   end
```


What do you think about this?

bbatsov replied:
Yeah, I think that's the best approach. I'm guessing that some cop about indentation consistency is flagging misindented comments even now, but if not - we should make sure those are flagged by some cop.

pocke commented:
I think `MaggicComment.parse` method should receive only comment line.  But current implementation receives not comment line. 

So I suggest the following patch instead.

```diff
diff --git a/lib/rubocop/cop/mixin/frozen_string_literal.rb b/lib/rubocop/cop/mixin/frozen_string_literal.rb
index 26b757a41..f863bd4c5 100644
--- a/lib/rubocop/cop/mixin/frozen_string_literal.rb
+++ b/lib/rubocop/cop/mixin/frozen_string_literal.rb
@@ -34,7 +34,7 @@ module RuboCop
       end
 
       def leading_comment_lines
-        processed_source[0..2].compact
+        processed_source.comments.select { |comment| comment.loc.line <= 3 }.map(&:text)
       end
     end
   end
```

What do you think?

koic commented:
@bbatsov @pocke Thanks for your feedback. To begin with, focusing on what to solve for issue #6115, I added some tests and fixed based on your advices.

rrosenblum commented on lib/rubocop/cop/mixin/frozen_string_literal.rb:
```diff
@@ -34,7 +34,9 @@ def frozen_string_literals_enabled?
 
       def leading_comment_lines
-        processed_source[0..2].compact
+        processed_source.comments.select do |comment|
```
We could use an `each_with_object` here to avoid chaining `map` onto `end`.

```ruby
processed_source.comments.each_with_object([]) do |comment, collection|
  next if comment.loc.line > 3
  collection << comment.text
end
```

koic replied:
Thanks for the review. I arranged it slightly and applied it.
